proxy/agent.go

```diff
@@ -0,0 +1,198 @@
+// limitations under the License.
+
+package proxy
+
+import (
+	"reflect"
+	"time"
+
+	"k8s.io/client-go/pkg/util/flowcontrol"
+
+	"github.com/golang/glog"
+)
+
+// Agent manages the restarts and the life cycle of a proxy binary.  Agent
+// keeps track of all running proxy processes and their configurations.  Hot
+// restarts are performed by launching a new proxy process with a strictly
+// incremented restart epoch.  This matches Envoy semantics for restart epochs:
+// To successfully launch a new Envoy process that will replace the running
+// Envoy processes, the restart epoch of the new process must be exactly 1
+// greater than the highest restart epoch of the currently running Envoy
+// processes. The initial epoch is 0.
+//
+// Envoy hot restart documentation:
+// https://lyft.github.io/envoy/docs/intro/arch_overview/hot_restart.html
+type Agent interface {
+	// Apply sets the desired configuration for the proxy.  Agent compares the
+	// current active configuration to the desired state and initiates a restart
+	// if necessary. If the restart fails, the agent attempts to retry with an
+	// exponential back-off.
+	Apply(config interface{})
+
+	// Run starts the agent and awaits for a signal on the input channel
+	Run(stop <-chan struct{})
+}
+
+// NewAgent creates a new proxy instance agent.  "run" function is a blocking
+// call to start the proxy. "cleanup" function is called after "run" finishes.
+// Both functions take the proxy  epoch as an argument. Retry is budgeted and
+// uses an exponential back-off from the initial delay.
+func NewAgent(run func(interface{}, int) error, cleanup func(int),
+	initialRetryBudget int, initialRetryDelay time.Duration) Agent {
+	return &agent{
+		run:                run,
+		cleanup:            cleanup,
+		epochs:             make(map[int]interface{}),
+		config:             make(chan interface{}),
+		status:             make(chan exitStatus),
+		initialRetryBudget: initialRetryBudget,
+		initialRetryDelay:  initialRetryDelay,
+	}
+}
+
+type agent struct {
+	// proxy start-up command
+	run func(interface{}, int) error
+
+	// proxy cleanup command
+	cleanup func(int)
+
+	// desired configuration state
+	desired interface{}
+
+	// active epochs and their configurations
+	epochs map[int]interface{}
+
+	// current configuration is the highest epoch configuration
+	current interface{}
+
+	// channel for posting desired configurations
+	config chan interface{}
+
+	// channel for proxy exit notifications
+	status chan exitStatus
+
+	// retry budget count
+	retryBudget int
+
+	// total budget available
+	initialRetryBudget int
+	initialRetryDelay  time.Duration
+}
+
+type exitStatus struct {
+	epoch int
+	err   error
+}
+
+func (a *agent) Apply(config interface{}) {
+	a.config <- config
+}
+
+// Run executes the reloads until stop channel closes
+func (a *agent) Run(stop <-chan struct{}) {
+	glog.V(2).Info("Starting proxy agent")
+
+	// Throttle processing up to smoothed 1 qps with bursts up to 10 qps
```

Comment:
Do we really need to update configuration 10 times per second? Do we have any idea of expectations for routing traffic after a new service is added, i.e. how fast do adopt new configuration?